File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        totalSum = 0
        pathSum = 0

        def dfs(node: Optional[TreeNode]):
            nonlocal pathSum
            nonlocal totalSum

            pathSum *= 10
            pathSum += node.val

            if node.left:
                dfs(node.left)

            if node.right:
                dfs(node.right)

            if not (node.left or node.right):
                totalSum += pathSum

            pathSum -= node.val
            pathSum //= 10

        dfs(root)

        return totalSum

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int totalSum = 0;
    int pathSum = 0;
    
    public int sumNumbers(TreeNode root) {
        
        dfs(root);
        return totalSum;
        
    }
    
    
    public void dfs(TreeNode node){
        pathSum *= 10;
        pathSum += node.val;

        if (node.left != null){
            dfs(node.left);
        }
        if (node.right != null){
            dfs(node.right);
        }
        
        if (node.left == null & node.right == null){
            totalSum += pathSum;
        }

        pathSum -= node.val;
        pathSum /= 10;
    }
}







